sunpy/net/dataretriever/sources/ace.py

```diff
@@ -0,0 +1,474 @@
+"""
+This module implements SWEPAM, MAG, SIS and EPAM Clients.
+"""
+
+import os
+import datetime
+from urllib.parse import urlsplit
+
+import astropy.units as u
+from astropy.time import Time, TimeDelta
+
+from sunpy.time import TimeRange
+from sunpy.net.dataretriever.client import GenericClient
+
+__all__ = ['SWEPAMClient', 'EPAMClient', 'MAGClient', 'SISClient']
+
+
+class SWEPAMClient(GenericClient):
+    """
+    Returns a list of URLS to ACE SWEPAM files corresponding to value of input timerange.
+    URL source: `ftp://ftp.swpc.noaa.gov/pub/lists/ace/`.
+    The earliest date available is from 29th July 2015.
+
+    Parameters
+    ----------
+    timerange : `sunpy.time.TimeRange`
+        Time range for which data is to be downloaded.
+
+    Returns
+    -------
+    `list`
+          List of urls corresponding to requested time range.
+
+    Examples
+    --------
+    >>> from sunpy.net import Fido
+    >>> from sunpy.net import attrs as a
+    >>> results = Fido.search(a.Time('2016/5/18 00:00:00', '2016/5/20 00:03:00'),
+    ...                       a.Instrument('swepam'))   #doctest: +REMOTE_DATA
+    >>> print(results)  #doctest: +REMOTE_DATA
+    Results from 1 Provider:
+    <BLANKLINE>
+    3 Results from the SWEPAMClient:
+         Start Time           End Time      Source Instrument Wavelength
+           str19               str19         str3     str6       str3
+    ------------------- ------------------- ------ ---------- ----------
+    2016-05-18 00:00:00 2016-05-18 23:59:59    ACE     swepam        nan
+    2016-05-19 00:00:00 2016-05-19 23:59:59    ACE     swepam        nan
+    2016-05-20 00:00:00 2016-05-20 23:59:59    ACE     swepam        nan
+    <BLANKLINE>
+    <BLANKLINE>
+    >>> response = Fido.fetch(results)  #doctest: +SKIP
+    """
+
+    def _get_time_for_url(self, urls):
+        times = []
+        for uri in urls:
+            uripath = urlsplit(uri).path
+
+            # Extract the yymmdd or yyyymmdd timestamp
+            datestamp = os.path.splitext(os.path.split(uripath)[1])[0][:8]
+
+            start = Time.strptime(datestamp, "%Y%m%d")
+            almost_day = TimeDelta(1 * u.day - 1 * u.millisecond)
+            times.append(TimeRange(start, start + almost_day))
+
+        return times
+
+    def _get_url_for_timerange(self, timerange, **kwargs):
+        """
+        Return URL(s) for corresponding timerange.
+
+        Parameters
+        ----------
+        timerange : `~sunpy.time.TimeRange`
+            The time range you want the files for.
+
+        Returns
+        -------
+        `list`
+            The URL(s) for the corresponding timerange.
+        """
+        START_DATE = datetime.datetime(2015, 7, 29)
+        if timerange.start < START_DATE:
+            raise ValueError(
+                'Earliest date for which SWEPAM data is available is '
+                '{:%Y-%m-%d}'.format(START_DATE))
+        base_url = 'ftp://ftp.swpc.noaa.gov/pub/lists/ace/'
```

Comment:
Can we / should we not use the [Scraper](https://docs.sunpy.org/en/stable/api/sunpy.util.scraper.Scraper.html#sunpy.util.scraper.Scraper) to get these files?

It's possible when this code was written the first time it didn't exist. We are generally trying to move away from simple pattern based URL generation.

Reply:
I forget we even have that.

Reply:
Yes, I even thought of implementing the scraper, it will make the url formation easy and more understandable. Doing it for all clients.

Reply:
@Cadair @nabobalis I have made changes, please review it.